pkg/backends/calico/routes.go

```diff
@@ -14,31 +14,265 @@
+func (rg *routeGenerator) onSvcDelete(obj interface{}) {
+	log.Info("+onSvcDelete")
+	rg.Lock()
+	defer rg.Unlock()
+
+	svc := obj.(*v1.Service)
+	delete(rg.svcEpMap, hash(&svc.ObjectMeta))
+	rg.updateRoutes()
+}
+
+// waitUntilExists continuously sleeps until the object is found in the svcEpMap
+func (rg *routeGenerator) waitUntilExists(key string) {
+	for {
+		if _, exists := rg.svcEpMap[key]; exists {
+			break
+		}
+		rg.Unlock()
+		time.Sleep(time.Second)
+		rg.Lock()
+		continue
+	}
+}
+
+// onEPAdd is called when a k8s endpoint is created
+// - adds the endpointIPs to the corresponding serviceEndpoints object in the map
+// - sleeps for one second if the serviceEndpoint object does not yet exist
+func (rg *routeGenerator) onEPAdd(obj interface{}) {
+	log.Info("+onEPAdd")
+	ep := obj.(*v1.Endpoints)
+	rg.Lock()
+	defer rg.Unlock()
+
+	// wait until the serviceEndpoint obj exists
+	svcEpKey := hash(&ep.ObjectMeta)
+	rg.waitUntilExists(svcEpKey)
+
+	log.WithFields(log.Fields{
+		"key": svcEpKey,
+	}).Info("onEPAdd")
+
+	// iterate over the subsets and addresses to add the endpoints
+	// that are on this node
+	for _, subset := range ep.Subsets {
+		for _, address := range subset.Addresses {
+			log.WithFields(log.Fields{
+				"addr": address.Hostname,
+			}).Info("onEPAdd")
+			if address.NodeName == nil || *address.NodeName != rg.nodeName {
+				continue
+			}
+			log.Info("onEPAdd: added")
+			rg.svcEpMap[svcEpKey].endpointIPs[address.Hostname] = true
+		}
+	}
+	rg.updateRoutes()
+}
+
+// onEPUpdate ...
+func (rg *routeGenerator) onEPUpdate(oldObj, newObj interface{}) {
+	oldEP, newEP := oldObj.(*v1.Endpoints), newObj.(*v1.Endpoints)
+	oldEPKey, newEPKey := hash(&oldEP.ObjectMeta), hash(&newEP.ObjectMeta)
+	rg.Lock()
+	defer rg.Unlock()
+
+	// iterate over old subsets and addresses to remove those ips
+	for _, subset := range oldEP.Subsets {
+		for _, address := range subset.Addresses {
+			if address.NodeName == nil || *address.NodeName != rg.nodeName {
+				continue
+			}
+			delete(rg.svcEpMap[oldEPKey].endpointIPs, address.Hostname)
+		}
+	}
+
+	// iterate over new subsets and addresses to add those ips
+	for _, subset := range newEP.Subsets {
+		for _, address := range subset.Addresses {
+			if address.NodeName == nil || *address.NodeName != rg.nodeName {
+				continue
+			}
+			rg.svcEpMap[newEPKey].endpointIPs[address.Hostname] = true
+		}
+	}
+	rg.updateRoutes()
+}
+
+// onEPDelete ...
+func (rg *routeGenerator) onEPDelete(obj interface{}) {
+	ep := obj.(*v1.Endpoints)
+	rg.Lock()
+	defer rg.Unlock()
+
+	// check that the svcEp obj isn't already gone
+	key := hash(&ep.ObjectMeta)
+	if _, exists := rg.svcEpMap[key]; !exists {
+		return
+	}
+
+	// iterate over subsets and addresses to remove the endpoints
+	for _, subset := range ep.Subsets {
+		for _, address := range subset.Addresses {
+			if address.NodeName == nil || *address.NodeName != rg.nodeName {
+				continue
+			}
+			delete(rg.svcEpMap[key].endpointIPs, address.Hostname)
```

Comment:
Perhaps I've misunderstood `Hostname`, but I assume it would be the same for all Endpoints on the same node.  Therefore I think this would leave us in the wrong state:
1. Service created.
2. Endpoints A and B created, both on this node: endpointIPs[hostname] = true
3. Endpoint B deleted: endpointIPs[hostname] is deleted.

Reply:
@neiljerram Hostname is the name of the pod, not the pod's host.

NodeName is the pod's host. It's kinda confusing :/